src/source/Ice/NatBehaviorDiscovery.c

```diff
@@ -0,0 +1,404 @@
+
+    if (bindingResponse == NULL) {
+        natMappingBehavior = NAT_BEHAVIOR_NO_UDP_CONNECTIVITY;
+        CHK(FALSE, retStatus);
+    }
+
+    CHK_STATUS(getMappAddressAttribute(bindingResponse, &pStunAttributeMappedAddress));
+    CHK_STATUS(getStunAttribute(bindingResponse,
+                                STUN_ATTRIBUTE_TYPE_CHANGED_ADDRESS,
+                                (PStunAttributeHeader *) &pStunAttributeOtherAddress));
+    CHK_ERR(pStunAttributeOtherAddress != NULL, retStatus, "Expect binding response to have other address or changed address attribute");
+    mappedAddress = pStunAttributeMappedAddress->address;
+    otherAddress = pStunAttributeOtherAddress->address;
+
+    if (isSameIpAddress(&pSocketConnection->hostIpAddr, &pStunAttributeMappedAddress->address, TRUE)) {
+        natMappingBehavior = NAT_BEHAVIOR_NOT_BEHIND_ANY_NAT;
+        CHK(FALSE, retStatus);
+    }
+
+    /* execute test II */
+    DLOGD("Running mapping behavior test II. Send binding request to alternate address but primary port");
+    testDestAddress = otherAddress;
+    testDestAddress.port = pStunServer->ipAddress.port;
+    CHK_STATUS(executeNatTest(bindingRequest, &testDestAddress, pSocketConnection, testIndex++, data, &bindingResponse));
+    CHK_ERR(bindingResponse != NULL, retStatus, "Expect to receive binding response");
+
+    CHK_STATUS(getMappAddressAttribute(bindingResponse, &pStunAttributeMappedAddress));
+
+    /* compare mapped address from test I and test II */
+    if (isSameIpAddress(&mappedAddress, &pStunAttributeMappedAddress->address, TRUE)) {
+        natMappingBehavior = NAT_BEHAVIOR_ENDPOINT_INDEPENDENT;
+        CHK(FALSE, retStatus);
+    }
+    mappedAddress = pStunAttributeMappedAddress->address;
+
+    /* execute test III */
+    DLOGD("Running mapping behavior test III. Send binding request to alternate address");
+    CHK_STATUS(executeNatTest(bindingRequest, &otherAddress, pSocketConnection, testIndex++, data, &bindingResponse));
+    CHK_ERR(bindingResponse != NULL, retStatus, "Expect to receive binding response");
+
+    CHK_STATUS(getMappAddressAttribute(bindingResponse, &pStunAttributeMappedAddress));
+    /* compare mapped address from test II and test III */
+    if (isSameIpAddress(&mappedAddress, &pStunAttributeMappedAddress->address, TRUE)) {
+        natMappingBehavior = NAT_BEHAVIOR_ADDRESS_DEPENDENT;
+    } else {
+        natMappingBehavior = NAT_BEHAVIOR_PORT_DEPENDENT;
+    }
+
+CleanUp:
+
+    if (pNatMappingBehavior != NULL) {
+        *pNatMappingBehavior = natMappingBehavior;
+    }
+
+    if (bindingRequest != NULL) {
+        freeStunPacket(&bindingRequest);
+    }
+
+    if (data != NULL) {
+        for (i = 0; i < data->bindingResponseCount; ++i) {
+            freeStunPacket(&data->bindingResponseList[i]);
+        }
+    }
+
+    CHK_LOG_ERR(retStatus);
+    return retStatus;
+}
+
+STATUS natFilteringBehaviorTest(PIceServer pStunServer, PNatTestData data, PSocketConnection pSocketConnection,
```

Comment:
same question about the prefix

Reply:
this mirrors the "test" described in the rfc: https://tools.ietf.org/html/rfc5780#section-4.3